firestore/src/Firestore.gs

```diff
@@ -0,0 +1,272 @@
+    var values = [];
+    // Provide values in the order defined by the schema.
+    schema.forEach(function(field) {
+      if (instance.PREDEFINED_FIELDS.indexOf(field.name) >= 0) {
+        var value = document[field.name];
+        switch(field.name) {
+          case 'createTime':
+          case 'updateTime':
+            values.push(instance.parseTimestamp(value));
+            break;
+          case 'name':
+            values.push(instance.parseDocumentId(value));
+            break;
+          default:
+            values.push(value);
+        }
+      } else {
+        const valueWrapper = document.fields[field.name];
+        if (valueWrapper) {
+          switch(field.dataType) {
+            case 'STRING':
+              if (field.semantics && field.semantics.semanticGroup == 'DATE_AND_TIME') {
+                values.push(instance.parseTimestamp(valueWrapper.timestampValue)); 
+              } else {
+                values.push(valueWrapper.stringValue);
+              }
+              break;
+            case 'NUMBER':
+              if (valueWrapper.integerValue != null) {
+                values.push(valueWrapper.integerValue);
+              } else {
+                values.push(valueWrapper.doubleValue);
+              }
+              break;
+            case 'BOOLEAN':
+              values.push(valueWrapper.booleanValue);
+              break;
+            default:
+              throw 'Unsupported data type: ' + field.dataType;    
+          }
+        } else {
+          values.push(null); 
+        }
+      }
+    });
+    data.push({
+      values: values
+    });
+  });
+  
+  return data;
+}
+
+
+/**
+ * Fetches documents from Firestore or cache.
+ *
+ * @param {string} project The Google Cloud project ID containing Firestore instance.
+ * @param {string} collection The Firestore collection to use.
+ * @param {number} numResults Maximum documents to fetch.
+ * @return {Array} Array of Firestore documents.
+ */
+Firestore.prototype.fetchDocuments = function(project, collection, numResults) {
+  const urlComponents = [
+    'https://firestore.googleapis.com/v1beta1/projects/',
+    project,
+    '/databases/(default)/documents/',
+    collection,
+    '?pageSize=',
+    this.PAGE_SIZE];
+  const url = urlComponents.join('');
+  
+  const cached = this.cache.get(url);
+  if (cached) {
+    return JSON.parse(cached); 
+  }
+  
+  var documents = [];
+  var token = null;
+  while (documents.length < numResults) {
+    var response = this.fetchPage(url, token);
+    if (response.documents && response.documents.length > 0) {
+      Array.prototype.push.apply(documents, response.documents);
+    }
+    if (response.nextPageToken && token != response.nextPageToken) {
+      token = response.nextPageToken;
+    } else {
+      break; 
+    }
+  }
+  
+  try {
+    this.cache.put(url, JSON.stringify(documents), this.CACHE_TTL); 
+  } catch (e) {
+    // Ignore. This usually means the data is too big to cache (which does make 
```

Comment:
You can also zip the data to make it more likely to fit in cache. See [zipBlobs](https://developers.google.com/apps-script/reference/utilities/utilities#zipblobs)

Reply:
Done